tokenclient/tokenFlows.go

```diff
@@ -0,0 +1,177 @@
+		if options.TLSConfig == nil && identity.IsCertificateBased() {
+			defaultConfig, err := defaultTLSConfig(identity)
+			if err != nil {
+				return nil, err
+			}
+			options.TLSConfig = defaultConfig
+		}
+		options.HTTPClient = defaultHTTPClient(options.TLSConfig)
+	}
+	t.options = options
+	t.tokenURI = identity.GetURL() + tokenEndpoint
+	return t, nil
+}
+
+// ClientCredentials implements the client credentials flow (RFC 6749, section 4.4).
+// Clients obtain an access token outside of the context of a user.
+// It is used for non interactive applications (a CLI, a batch job, or for service-2-service communication) where the token is issued to the application itself,
+// instead of an end user for accessing resources without principal propagation.
+//
+// ctx carries the request context like the deadline or other values that should be shared across API boundaries. Default: context.TODO is used
+// customerTenantURL like "https://custom.accounts400.ondemand.com" gives the host of the customers ias tenant
+// options allows to provide a request context and optionally additional request parameters
+func (t *TokenFlows) ClientCredentials(ctx context.Context, customerTenantURL string, options RequestOptions) (string, error) {
+	data := url.Values{}
+	data.Set(grantTypeParameter, grantTypeClientCredentials)
+	data.Set(clientIDParameter, t.identity.GetClientID())
+	if t.identity.GetClientSecret() != "" {
+		data.Set(clientSecretParameter, t.identity.GetClientSecret())
+	}
+	for name, value := range options.Params {
+		data.Set(name, value) // potentially overwrites data which was set before
+	}
+	targetURL, err := t.getURL(customerTenantURL)
+	if err != nil {
+		return "", err
+	}
+	r, e := http.NewRequestWithContext(ctx, http.MethodPost, targetURL, strings.NewReader(data.Encode())) // URL-encoded payload
+	if e != nil {
+		return "", fmt.Errorf("error performing client credentials flow: %w", e)
+	}
+	r.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	tokenJSON, err := t.performRequest(r)
+	if err != nil {
+		return "", err
+	}
+	var response tokenResponse
+	_ = json.Unmarshal(tokenJSON, &response)
+	if response.Token == "" {
+		return "", fmt.Errorf("error parsing requested client credential token: %v", string(tokenJSON))
+	}
+	return response.Token, nil
+}
+
+func (t *TokenFlows) getURL(customerTenantHost string) (string, error) {
+	if customerTenantHost == "" {
+		return t.tokenURI, nil
+	}
+	customHost, err := url.Parse(customerTenantHost)
+	if err == nil && customHost.Host != "" {
+		return "https://" + customHost.Host + tokenEndpoint, nil
+	}
+	return "", fmt.Errorf("customer tenant host '%v' can't be accepted: %v", customerTenantHost, err)
+}
+
+func (t *TokenFlows) performRequest(r *http.Request) ([]byte, error) {
+	res, err := t.options.HTTPClient.Do(r)
+	if err != nil {
+		return nil, fmt.Errorf("request to '%v' failed: %w", r.URL, err)
+	}
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
```

Comment:
cool - thanks!

Reply:
After analysis, the issue with using decoder was that i wanted to separate the request and error handling from json parsing into the respective struct. For now, its always the same, so this is still possible by refactoring the `performRequest` internal func. Otherwise it wouldn't be possible as the io connection is only accessible within the func scope and afterwards deferred.